firmware/shared/code/CAN.c

```diff
@@ -0,0 +1,313 @@
+	GPIO_InitStructure.GPIO_Speed = GPIO_Speed_50MHz;
+	GPIO_InitStructure.GPIO_OType = GPIO_OType_PP;
+	GPIO_InitStructure.GPIO_PuPd = GPIO_PuPd_UP;
+	GPIO_Init(CAN_config.HWConfig->GPIOPort, &GPIO_InitStructure);
+
+	// Attach GPIO AF to UART
+	GPIO_PinAFConfig(CAN_config.HWConfig->GPIOPort, CAN_config.HWConfig->rxPin, CAN_config.HWConfig->AFNumber);
+	GPIO_PinAFConfig(CAN_config.HWConfig->GPIOPort, CAN_config.HWConfig->txPin, CAN_config.HWConfig->AFNumber);
+}
+
+static void CAN_private_CANPeriphInit(void)
+{
+    assert(IS_CAN_ALL_PERIPH(CAN_config.HWConfig->CANPeriph));
+
+    CAN_InitTypeDef CANInitStruct;
+    CAN_StructInit(&CANInitStruct);
+
+    // Sample point at 90%
+    CANInitStruct.CAN_Prescaler = PCLK / (BAUD_RATE * NUMBER_OF_TIME_QUANTA); // Specifies the length of a time quantum
+    
+    // The parameters below are described in the CAN standard
+    CANInitStruct.CAN_SJW = CAN_SJW_1tq; // Specifies the maximum number of time quanta the CAN hardware is allowed to lengthen or shorten a bit to perform resynchronization
+    CANInitStruct.CAN_BS1 = CAN_BS1_8tq; // Specifies the number of time quanta in Bit Segment 1
+    CANInitStruct.CAN_BS2 = CAN_BS2_1tq; // Specifies the number of time quanta in Bit Segment 2
+
+    // CANInitStruct.CAN_Mode = CAN_Mode_Normal;
+    CANInitStruct.CAN_Mode = CAN_Mode_Silent_LoopBack; // Used for testing
+    CANInitStruct.CAN_TTCM = DISABLE; // Time triggered communication mode
+    CANInitStruct.CAN_ABOM = DISABLE; // Automatic bus-off management
+    CANInitStruct.CAN_AWUM = DISABLE; // Automatic wake-up mode
+    CANInitStruct.CAN_NART = DISABLE; // Non-automatic retransmission mode
+    CANInitStruct.CAN_RFLM = DISABLE; // Receive FIFO Locked mode
+    CANInitStruct.CAN_TXFP = ENABLE; // Transmit FIFO priority
+
+    if(CAN_Init(CAN_config.HWConfig->CANPeriph, &CANInitStruct) != CAN_InitStatus_Success)
+    {
+        debug_writeStringBlocking("CAN init failed");
+        CAN_data.initSuccessful = false;
+    }
+
+    // Hard coded to only use FIFO_0
+    // Enable FIFO 0 not empty interrupt
+    CAN_ITConfig(CAN_config.HWConfig->CANPeriph, CAN_IT_FMP0, ENABLE);
+	NVIC_SetPriority(CAN_private_getIRQNumber(CAN_config.HWConfig->CANPeriph), 5);
+	NVIC_EnableIRQ(CAN_private_getIRQNumber(CAN_config.HWConfig->CANPeriph));
+}
+
+static uint32_t CAN_private_getIRQNumber(const CAN_TypeDef * periph)
+{
+    uint32_t ret;
+
+    switch((uint32_t)periph)
+    {
+        case (uint32_t)CAN1:
+        {
+            ret = CAN_interruptNumberMaping[CAN_PERIPH_1];
+            break;
+        }                
+        case (uint32_t)CAN2:
+        {
+            ret = CAN_interruptNumberMaping[CAN_PERIPH_2];
+            break;
+        }            
+        case (uint32_t)CAN3:
+        {
+            ret = CAN_interruptNumberMaping[CAN_PERIPH_3];
+            break;
+        }            
+        default:
+        {
+            ret = 500; // Unreasonably large number

```

Comment:
Why cant we return NULL instead of a random large number?

Reply:
`ret` is not a pointer. Setting it to NULL would just assign it 0 which is a valid number.